packages/sync/src/firehose/index.ts

```diff
@@ -0,0 +1,342 @@
+        )
+        return filtered
+      } else if (isAccount(evt) && !this.opts.excludeAccount) {
+        const parsed = parseAccount(evt)
+        return parsed ? [parsed] : []
+      } else if (isIdentity(evt) && !this.opts.excludeIdentity) {
+        const parsed = await parseIdentity(
+          this.opts.idResolver,
+          evt,
+          this.opts.unauthenticatedHandles,
+        )
+        return parsed ? [parsed] : []
+      } else {
+        return []
+      }
+    } catch (err) {
+      this.opts.onError?.(new FirehoseParseError(err, evt))
+      return []
+    }
+  }
+
+  private async processEvt(evt: RepoEvent) {
+    const parsed = await this.parseEvt(evt)
+    for (const write of parsed) {
+      try {
+        await this.opts.handleEvt(write)
+      } catch (err) {
+        this.opts.onError(new FirehoseHandlerError(err, write))
+      }
+    }
+  }
+
+  async destroy(): Promise<void> {
+    this.abortController.abort()
+    await this.destoryDefer.complete
+  }
+}
+
+export const parseCommitAuthenticated = async (
+  idResolver: IdResolver,
+  evt: Commit,
+  forceKeyRefresh = false,
+): Promise<CommitEvt[]> => {
+  const did = evt.repo
+  const key = await idResolver.did.resolveAtprotoKey(did, forceKeyRefresh)
+  const claims = evt.ops.map((op) => {
+    const { collection, rkey } = parseDataKey(op.path)
+    return {
+      collection,
+      rkey,
+      cid: op.action === 'delete' ? null : op.cid,
+    }
+  })
+  const verifiedCids: Record<string, CID | null> = {}
+  try {
+    const results = await verifyProofs(evt.blocks, claims, did, key)
+    results.verified.forEach((op) => {
+      const path = formatDataKey(op.collection, op.rkey)
+      verifiedCids[path] = op.cid
+    })
+  } catch (err) {
+    if (err instanceof RepoVerificationError) {
+      return parseCommitAuthenticated(idResolver, evt, true)
+    }
+    throw err
+  }
+  const verifiedOps: RepoOp[] = evt.ops.filter((op) => {
+    if (op.action === 'delete') {
+      return verifiedCids[op.path] === null
+    } else {
+      return op.cid !== null && op.cid.equals(verifiedCids[op.path])
+    }
+  })
+  return formatCommitOps(evt, verifiedOps)
+}
+
+export const parseCommitUnauthenticated = async (
+  evt: Commit,
+): Promise<CommitEvt[]> => {
+  return formatCommitOps(evt, evt.ops)
+}
+
+const formatCommitOps = async (evt: Commit, ops: RepoOp[]) => {
+  const car = await readCar(evt.blocks)
+
+  const evts: CommitEvt[] = []
+
+  for (const op of ops) {
+    const uri = new AtUri(`at://${evt.repo}/${op.path}`)
+
+    const meta: CommitMeta = {
+      seq: evt.seq,
+      time: evt.time,
+      commit: evt.commit,
+      blocks: car.blocks,
+      rev: evt.rev,
+      uri,
+      did: uri.host,
+      collection: uri.collection,
+      rkey: uri.rkey,
+    }
+
+    if (op.action === 'create' || op.action === 'update') {
+      if (!op.cid) continue
+      const recordBytes = car.blocks.get(op.cid)
+      if (!recordBytes) continue
```

Comment:
I wonder if there's a case that these should cause an error in the `parseCommitAuthenticated()` path.